src/brod_app.erl

```diff
@@ -0,0 +1,56 @@
+%%%
+%%%   Copyright (c) 2015 Klarna AB
+%%%
+%%%   Licensed under the Apache License, Version 2.0 (the "License");
+%%%   you may not use this file except in compliance with the License.
+%%%   You may obtain a copy of the License at
+%%%
+%%%       http://www.apache.org/licenses/LICENSE-2.0
+%%%
+%%%   Unless required by applicable law or agreed to in writing, software
+%%%   distributed under the License is distributed on an "AS IS" BASIS,
+%%%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%%%   See the License for the specific language governing permissions and
+%%%   limitations under the License.
+%%%
+
+%%%=============================================================================
+%%% @doc
+%%% @copyright 2015 Klarna AB
+%%% @end
+%%%=============================================================================
+-module(brod_app).
+-behaviour(application).
+
+-export([ start/2
+        , start/0
+        , stop/1
+        ]).
+
+%% application callback
+start(_StartType, _StartArgs) -> start().
+
+%% @doc Start brod application.
+%%
+%% Permanent producers can be configured in app env (sys.config).
+%% A minimal example of app env with permanent producer args:
+%%    [ { producers
+%%      , [ {producer_id_1, [{hosts, [{"localhost", 9092}]}]}
+%%        , {producer_id_2, ...}
+%%        ]
+%%      }
+%%    ].
+%% @see brod:start_link_producer/2 for more info about all producer args
+%%
+%% @end
+start() ->
+  _ = application:load(brod), %% ensure loaded
+  %% if no producer is configured in app env, start the supervisor empty
+  PermanentProducers = application:get_env(brod, producers, _Default = []),
```

Comment:
I think the relation between `start/0` and `start/2` are usually the other way around:

* `start/2` is the application controller's callback function, and starts the supervision tree on its own. It can assume the application is loaded of course, so it can access the env variables.
* `start/0` is a convenience wrapper, and calls `application:ensure_[all_]started/1`.

The problem with this code is that `start/0` will start the supervision tree linked to the calling process (not what you'd expect), without starting the application (so subsequent attempts to start `brod` as an application will fail).